sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,769 @@
+                                                self.significant_min,
+                                                self.min_cluster_size_ratio,
+                                                self.min_maxima_ratio)
+        self.core_sample_indices_ = indices_
+        self.n_clusters_ = np.max(self.labels_)
+        return self
+
+    # OPTICS helper functions; these should not be public #
+
+    def _expand_cluster_order(self, point, X):
+        # As above, not parallelizable. Parallelizing would allow items in
+        # the 'unprocessed' list to switch to 'processed'
+        if self.core_dists_[point] <= self.max_bound:
+            while not self._processed[point]:
+                self._processed[point] = True
+                self.ordering_.append(point)
+                point = self._set_reach_dist(point, X)
+        else:
+            self.ordering_.append(point)  # For very noisy points
+            self._processed[point] = True
+
+    def _set_reach_dist(self, point_index, X):
+        P = np.array(X[point_index]).reshape(1, -1)
+        indices = self.radius_neighbors(P, radius=self.max_bound,
+                                        return_distance=False)[0]
+
+        # Checks to see if there more than one member in the neighborhood
+        if iterable(indices):
+            # Masking processed values; n_pr is 'not processed'
+            n_pr = np.compress((np.take(self._processed,
+                                        indices, axis=0) < 1).ravel(),
+                               indices, axis=0)
+            # n_pr = indices[(self._processed[indices] < 1).ravel()]
+            # Keep n_jobs = 1 in the following lines...please
+            if len(n_pr) > 0:
+                dists = pairwise_distances(P, np.take(X, n_pr, axis=0),
+                                           self.metric, n_jobs=1).ravel()
+
+                rdists = np.maximum(dists, self.core_dists_[point_index])
+                new_reach = np.minimum(np.take(self.reachability_,
+                                               n_pr, axis=0), rdists)
+                self.reachability_[n_pr] = new_reach
+
+            # Checks to see if everything is already processed;
+            # if so, return control to main loop
+            if n_pr.size > 0:
+                # Define return order based on reachability distance
+                return(n_pr[quick_scan(np.take(self.reachability_,
+                                               n_pr, axis=0), dists)])
+            else:
+                return point_index
+
+    def extract_dbscan(self, eps):
+        """Performs DBSCAN extraction for an arbitrary epsilon.
+
+        Parameters
+        ----------
+        eps: float or int, required
+            DBSCAN `eps` parameter. Must be set to < `max_bound`. Equivalence
+            with DBSCAN algorithm is achieved if `eps` is < (`max_bound` * 5)
+
+        Returns
+        -------
+        core_sample_indices_: array, shape (n_core_samples,)
```

Comment:
space before :

Reply:
ok, I assume the same style requirement on 'Returns' applies to 'Attributes' and 'Parameters' too, so I looked through the rest of the docstrings and added spaces where needed.